src/UglyToad.PdfPig/Writer/PdfMerger.cs

```diff
@@ -415,15 +415,15 @@ private IToken CopyToken(IToken tokenToCopy, IPdfTokenScanner tokenScanner, IDic
                                 return newReferenceToken;
                             }
 
-                            var tokenObject = DirectObjectFinder.Get<IToken>(referenceToken.Data, tokenScanner);
+                            //we add the token to referencesFromDocument to prevent stackoverflow on references cycles 
+                            newReferenceToken = context.WriteToken(null);
```

Comment:
You can reserve a reference by doing `PdfStreamWriter::ReserveNumberToken`, by doing that you don't have to write a null token. I was expecting that `PdfStreamWriter::WriteToken` should not let you write null token, because this would be incompatible with an immediate flusher pdf stream writer
```suggestion
                            //we add the token to referencesFromDocument to prevent stackoverflow on references cycles 
                            newReferenceToken = context.ReserveNumberToken();
                            referencesFromDocument.Add(referenceToken, newReferenceToken);

                            var tokenObject = DirectObjectFinder.Get<IToken>(referenceToken.Data, tokenScanner);
                            Debug.Assert(!(tokenObject is IndirectReferenceToken));

                            var newToken = CopyToken(tokenObject, tokenScanner, referencesFromDocument);
                            return context.WriteToken(newToken, (int)newReferenceToken.Data.ObjectNumber);
```

Reply:
thanks. i updated the branch